File: backend/app/schemas/video.py
"""Pydantic schemas for video API responses."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class VideoRead(BaseModel):
    """Public video shape for list, search, and semantic endpoints."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    creator_name: str
    channel_url: str
    video_url: str = ""
    subscribers_count: int
    title: str
    views_count: int
    published_at: datetime | None
    created_at: datetime

    has_transcript: bool = False
    transcript_preview: str | None = None

    # Semantic search scores (pgvector)
    similarity_score: float | None = None
    title_similarity: float | None = None
    transcript_similarity: float | None = None
    match_source: str | None = None  # title | transcript | both | keyword

    # LangGraph retrieval snippet
    transcript_snippet: str | None = None
    comment_snippet: str | None = None


class VideoDetail(VideoRead):
    """Single video — includes full transcript."""

    transcript: str | None = None


class CatalogStats(BaseModel):
    """Embedding coverage for runtime-aware UI copy."""

    video_count: int
    title_embedding_count: int
    transcript_embedding_count: int


class VideoListResponse(BaseModel):
    videos: list[VideoRead]
    total: int


class SyncResult(BaseModel):
    created: int
    updated: int
    total_rows: int
    embeddings_created: int = 0
    transcripts_fetched: int = 0
    transcript_embeddings_created: int = 0
    hooks_indexed: int = 0
    comments_fetched: int = 0

File: frontend/types/video.ts
/** Video entity — mirrors backend VideoRead / VideoDetail */

export interface Video {
  id: number;
  creator_name: string;
  channel_url: string;
  video_url?: string;
  subscribers_count: number;
  title: string;
  views_count: number;
  published_at: string | null;
  created_at: string;
  has_transcript: boolean;
  transcript_preview?: string | null;
  transcript?: string | null;
  similarity_score?: number | null;
  title_similarity?: number | null;
  transcript_similarity?: number | null;
  match_source?: string | null;
  transcript_snippet?: string | null;
}

export interface CatalogStats {
  video_count: number;
  title_embedding_count: number;
  transcript_embedding_count: number;
}

export interface VideoListResponse {
  videos: Video[];
  total: number;
}

export interface SyncResult {
  created: number;
  updated: number;
  total_rows: number;
  embeddings_created: number;
  transcripts_fetched: number;
  transcript_embeddings_created: number;
  hooks_indexed: number;
  comments_fetched?: number;
}
